make/jdk/src/classes/build/tools/taglet/SealedGraph.java

```diff
@@ -0,0 +1,259 @@
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package build.tools.taglet;
+
+import com.sun.source.doctree.DocTree;
+import jdk.javadoc.doclet.Doclet;
+import jdk.javadoc.doclet.DocletEnvironment;
+import jdk.javadoc.doclet.Taglet;
+
+import javax.lang.model.element.*;
+import javax.lang.model.type.DeclaredType;
+
+import java.io.*;
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static java.lang.System.lineSeparator;
+import static jdk.javadoc.doclet.Taglet.Location.TYPE;
+
+/**
+ * A block tag to optionally insert a reference to a sealed class hierarchy graph,
+ * and generate the corresponding dot file.
+ */
+public final class SealedGraph implements Taglet {
+    private static final String sealedGraphDotPath =
+            System.getProperty("sealedGraphDotPath");
+
+    private DocletEnvironment docletEnvironment;
+
+    /**
+     * Returns the set of locations in which a taglet may be used.
+     */
+    @Override
+    public Set<Location> getAllowedLocations() {
+        return EnumSet.of(TYPE);
+    }
+
+    @Override
+    public boolean isInlineTag() {
+        return false;
+    }
+
+    @Override
+    public String getName() {
+        return "sealedGraph";
+    }
+
+    @Override
+    public void init(DocletEnvironment env, Doclet doclet) {
+        docletEnvironment = env;
+    }
+
+    @Override
+    public String toString(List<? extends DocTree> tags, Element element) {
+        if (sealedGraphDotPath == null || sealedGraphDotPath.isEmpty()) {
+            return "";
+        }
+        if (docletEnvironment == null || !(element instanceof TypeElement typeElement)) {
+            return "";
+        }
+
+        ModuleElement module = docletEnvironment.getElementUtils().getModuleOf(element);
+        File dotFile = new File(sealedGraphDotPath,
+                module.getQualifiedName() + "_" + typeElement.getQualifiedName() + ".dot");
+
+        Set<String> exports = module.getDirectives().stream()
+                .filter(ModuleElement.ExportsDirective.class::isInstance)
+                .map(ModuleElement.ExportsDirective.class::cast)
+                // Only include packages that are globally exported (i.e. no "to" exports)
+                .filter(me -> me.getTargetModules() == null)
```

Comment:
The lambda parameter here is an `ExportsDirective`, not a `ModuleElement`, suggesting that the name here should be `ed` instead of `me`.

Reply:
Fixed.